README.md
README
================

<!-- README.md is generated from README.Rmd. Please edit that file -->

# BayesDIP

Provide early termination phase II trial designs with a decreasingly
informative prior (DIP) or a regular Bayesian prior chosen by the user.
The program can determine the minimum planned sample size necessary to
achieve the user-specified admissible designs. The program can also
perform power and expected sample size calculations for the tests in
early termination Phase II trials.\[1\]

## Installation

You can install from CRAN with:

``` r
install.packages("BayesDIP")
```

Or try the development version from \[GitHub\] with:

``` r
# install.packages("devtools")
devtools::install_github("chenw10/BayesDIP")
```

## Example

``` r
library(BayesDIP)

# Calculate the minimum planned sample size within the range 10<=N<=100,
# under an admissible design which is set as 80% power and 5% type I error here.

# One sample Bernoulli model with the response rate for the new treatment is 0.5, 
# the null response rate is 0.3, and the target improvement to achieve is 0.
# The alternative hypothesis: p1 > p0 + d
# Simulate 10 replicate trials using this design with efficacy boundary 0.98 
# and futility boundary 0.05.

### Designs with traditional Bayesian prior Beta(1,1)
### Designs and operating characteristics based on 100 simulations:
OneSampleBernoulli.Design(list(2,1,1), nmin = 10, nmax=100, p0 = 0.3, p1 = 0.5, d = 0,
                          ps = 0.98, pf = 0.02, power = 0.80, t1error=0.05, alternative = "greater",
                          seed = 202210, sim = 100)
#> 
#> Prior:   Beta(1,1) 
#> Planned Sample Size:   92 
#> Efficacy Boundary:   0.98 
#> Futility Boundary:   0.02 
#> Exact Power:  0.99 
#> Exact Type I error:   0.05 
#> Expected sample size:  24 
#> Expected sample size standard deviation:  17.6


### Designs with DIP
### Designs and operating characteristics based on 10 simulations:
OneSampleBernoulli.Design(list(1,0,0), nmin = 10, nmax=100, p0 = 0.3, p1 = 0.5, d = 0,
                          ps = 0.98, pf = 0.02, power = 0.80, t1error=0.05, alternative = "greater",
                          seed = 202210, sim = 100)
#> 
#> Prior:   DIP 
#> Planned Sample Size:   44 
#> Efficacy Boundary:   0.98 
#> Futility Boundary:   0.02 
#> Exact Power:  0.81 
#> Exact Type I error:   0.05 
#> Expected sample size:  30 
#> Expected sample size standard deviation:  9.7


# Calculate the power, type I error and the expected sample size given a planned sample size

# One sample Bernoulli model with the response rate for the new treatment is 0.5, 
# the null response rate is 0.3, and the target improvement to achieve is 0.05.
# The alternative hypothesis: p1 > p0 + d
# Simulate 100 replicate trials for a given planned sample size 100 using this design
# with efficacy boundary 0.98 and futility boundary 0.05.  

## with traditional Bayesian prior Beta(1,1)
## Operating characteristics based on 100 simulations:
OneSampleBernoulli(list(2,1,1), N = 100, p0 = 0.3, p1 = 0.5, d = 0.05,
                   ps = 0.98, pf = 0.05, alternative = "greater",
                   seed = 202210, sim = 100)
#> 
#> Prior:   Beta(1,1) 
#> Power:  0.89 
#> Type I error:   0.05 
#> Expected sample size:   42.7 
#> Expected sample size standard deviation:   29.06 
#> The probability of reaching the efficacy boundary:   0.89 
#> The probability of reaching the futility boundary:   0.02


## with DIP
## Operating characteristics based on 100 simulations:
OneSampleBernoulli(list(1,0,0), N = 100, p0 = 0.3, p1 = 0.5, d = 0.05,
                   ps = 0.98, pf = 0.05, alternative = "greater",
                   seed = 202210, sim = 100)
#> 
#> Prior:   DIP 
#> Power:  0.86 
#> Type I error:   0.01 
#> Expected sample size:   72.1 
#> Expected sample size standard deviation:   17.28 
#> The probability of reaching the efficacy boundary:   0.86 
#> The probability of reaching the futility boundary:   0
```

## Reference

\[1\] Wang C, Sabo RT, Mukhopadhyay ND, and Perera RA. Early termination
in single-parameter model phase II clinical trial designs using
decreasingly informative priors. , 9(2): April - June 2022.
<https://doi.org/10.18203/2349-3259.ijct20221110>


NEWS.md
# BayesDIP 0.1 (25 Jan 2023)

-----------------------------------------------------------

# BayesDIP 0.1.1 (31 Jan 2023)

* Add two references describing the methods in DESCRIPTION file.
* Reduce simulation number from "sim = 100" to "sim = 10" for the examples in functions ("OneSampleBernoulli", "OneSampleNormal1", "OneSampleNormal2", "OneSamplePoisson", "TwoSampleBernoulli"), and unwrap these examples in \donttest{}.
* Allow R generate objects that can be used to extract the information a user is interested in, and then print that object.
* Version numbers and dates updated.







R/OneSampleBernoulli.Design.R
# ------------------------------------------------------------------------------
#
#     prior --> list containing the information for prior
#          [[1]] - the prior distribution type:
#                1 - DIP
#                2 - Beta(a,b)
#          [[2]] - a: first parameter of the Beta distribution
#          [[3]] - b: second parameter of the Beta distribution
#
#
# ------------------------------------------------------------------------------

#' One sample Bernoulli model - Trial Design
#'
#' Calculate the minimum planned sample size under an admissible design.
#' The users decide the power and type-I-error, and pick the efficacy and futility boundaries.
#' If there are no admissible design based on controlled type-I-error, then default to output
#' the designs with the lowest type-I-error and at least the user-defined (e.g. 80\%) power.
#'
#'
#' @param prior A list of length 3 containing the distributional information of the prior.
#' The first element is a number specifying the type of prior. Options are
#' \enumerate{
#' \item DIP ;
#' \item Beta(a,b), where a = shape, b = scale}
#' The second and third elements of the list are the parameters a and b, respectively.
#' @param nmin The start searching sample size
#' @param nmax The stop searching sample size
#' @param p0 The null response rate, which could be taken as the standard or historical rate.
#' @param p1 The response rate of the new treatment.
#' @param d The target improvement (minimal clinically meaningful difference).
#' @param ps The efficacy boundary (upper boundary).
#' @param pf The futility boundary (lower boundary).
#' @param power The power to achieve.
#' @param t1error The controlled type-I-error.
#' @param alternative less (lower values imply greater efficacy) or greater (larger
#' values imply greater efficacy).
#' @param seed The seed for simulations.
#' @param sim The number of simulations.
#' @return A list of the arguments with method and computed elements.
#' @examples
#' \donttest{
#' # with traditional Bayesian prior Beta(1,1)
#' OneSampleBernoulli.Design(list(2,1,1), nmin = 10, nmax=100, p0 = 0.3, p1 = 0.5, d = 0,
#'                    ps = 0.98, pf = 0.02, power = 0.80, t1error=0.05, alternative = "greater",
#'                    seed = 202210, sim = 10)
#' # with DIP
#' OneSampleBernoulli.Design(list(1,0,0), nmin = 10, nmax=100, p0 = 0.3, p1 = 0.5, d = 0,
#'                    ps = 0.98, pf = 0.02, power = 0.80, t1error=0.05, alternative = "greater",
#'                    seed = 202210, sim = 10)
#' }
#' @importFrom stats rbeta rbinom rgamma rnorm rpois
#' @export OneSampleBernoulli.Design

OneSampleBernoulli.Design <- function(prior, nmin = 10, nmax = 100, p0, p1, d = 0,
                               ps, pf, power = 0.8, t1error = 0.05,
                               alternative = c("less", "greater"), seed = 202209, sim = 1000){

  alternative <- match.arg(alternative)
  # Define the inputs
  if(prior[[1]] == 1){
    prior[[2]] <- NA
    prior[[3]] <- NA
  }
  ## nmin limit
  if(!is.null(nmin) && (!is.numeric(nmin) || nmin < 10 || nmin >= nmax))
    stop("nmin must be positive number and at least 10")
  ## nmax limit
  if(!is.null(nmax) && (!is.numeric(nmax) || nmax <= nmin || nmax >= 200))
    stop("nmax must greater than 'nmin' and less than 200")
  ## p0 limit
  if(!is.null(p0) && (!is.numeric(p0) || (p0 < 0 | p0 > 1)))
    stop("p0 must be numeric in [0,1]")
  ## p1 limit
  if(!is.null(p1) && (!is.numeric(p1) || (p1 < 0 | p1 > 1)))
    stop("p1 must be numeric in [0,1]")
  ## d limit
  if(!is.null(d) && (!is.numeric(d) || (d < 0 | d > abs(p1-p0))))
    stop("d must be numeric in [0, |p1-p0|]")
  ## efficacy boundary limit
  if(!is.null(ps) && (!is.numeric(ps) || (ps < 0.8 | ps > 1)))
    stop("ps (efficacy boundary) must be numeric in [0.8,1]")
  ## futility boundary limit
  if(!is.null(pf) && (!is.numeric(pf) || (pf < 0 | pf > 0.2)))
    stop("pf (futility boundary) must be numeric in [0,0.2]")
  ## power limit
  if(!is.null(power) && (!is.numeric(power) || (power < 0 | power > 1)))
    stop("power must be numeric in [0,1]")
  ## t1error limit
  if(!is.null(t1error) && (!is.numeric(t1error) || (t1error < 0 |t1error > 1)))
    stop("type-I-error must be numeric in [0,1]")
  ## set.seed
  if(!is.numeric(seed))
    stop("seed must be numeric")
  if(!is.numeric(sim))
    stop("simulation number must be numeric")

  set.seed(seed)

  # Functions to calculate the posterior
  Bernoulli <- function(a,b,y){posterior<-rbeta(1000, a+sum(y), b+(length(y)-sum(y)))}
  Bernoulli.DIP <- function(p0, y, N){
    j<-length(y)
    posterior<-rbeta(1000,1+sum(y)+p0*(N-j),1+(j-sum(y))+(1-p0)*(N-j))
  }


# Simulated Data
# calculate N that can achieve the power
N_v <- NULL
power_v <- NULL
n_v <- NULL
sd_v <- NULL
for (N in seq(from=nmin, to=nmax, by=1)){
    cat1s <- 0
    cat1f <- 0
    n.enrolled <- NULL
    for (k in 1:sim) {
      y.data <- NULL
      j <- 0
      cat <- 0
      cats <- 0
      catf <- 0
      pp_stop <- 0.5
      while(cat == 0){
        j <- j+1
        y.data <- append(y.data, rbinom(1,1,p1))
        if(j>=10)
        {
          if (prior[[1]] == 2){
            p1_s<-Bernoulli(a = prior[[2]], b = prior[[3]], y = y.data)
          }else if (prior[[1]] == 1){
            p1_s <- Bernoulli.DIP(p0, y = y.data, N = N)
          }

          if (alternative == "greater"){
            pp_stop<-sum(p1_s>p0+d)/length(p1_s)
          }else if (alternative == "less"){
            pp_stop<-sum(p1_s<p0-d)/length(p1_s)
          }
        }
        if(pp_stop>=ps){cats<-1}
        if(pp_stop<pf){catf<-1}
        cat<-cats+catf
        if(j==N){cat<-1}
      }
      if(cats==1){cat1s<-cat1s+1}
      if(cats==0){cat1s<-cat1s}
      if(catf==1){cat1f<-cat1f+1}
      if(catf==0){cat1f<-cat1f}

      # Recruited Sample Size
      n.enrolled <- append(n.enrolled, j)
    }
    power.cal <- cat1s/sim

    jitter <- 0.01
    if (power.cal >= power-jitter){
    N_v <- append(N_v, N)
    power_v <- append(power_v, power.cal)
    n_v <- append(n_v, round(mean(n.enrolled), 0))
    sd_v <- append(sd_v, round(sd(n.enrolled), 1))
    }
    result1 <- cbind(N_v, power_v, n_v, sd_v)
} # End of power calculation


if (is.null(result1)){
  message("Suggest: please adjust your input values!")
  stop(paste("No sample size in the range [",nmin,",",nmax,"] can achieve ", power*100, "% power", sep=""))
}


# calculate type I error
nmin1 <- N_v[which.min(N_v)] # start minimum sample size in calculation of exact type I error
N_v <- NULL
t1error_v <- NULL
for (N in seq(from=nmin1, to=nmax, by=1)){
  cat1s <- 0
  cat1f <- 0
  for (k in 1:sim) {
    y.data <- NULL
    j <- 0
    cat <- 0
    cats <- 0
    catf <- 0
    pp_stop <- 0.5
    while(cat == 0){
      j <- j+1
      y.data <- append(y.data, rbinom(1,1,p0)) # under the null hypothesis p1 = p0
      if(j>=10)
      {
        if (prior[[1]] == 2){
          p1_s<-Bernoulli(a = prior[[2]], b = prior[[3]], y = y.data)
        }else if (prior[[1]] == 1){
          p1_s <- Bernoulli.DIP(p0, y = y.data, N = N)
        }

        if (alternative == "greater"){
          pp_stop<-sum(p1_s>p0+d)/length(p1_s)
        }else if (alternative == "less"){
          pp_stop<-sum(p1_s<p0-d)/length(p1_s)
        }
      }
      if(pp_stop>=ps){cats<-1}
      if(pp_stop<pf){catf<-1}
      cat<-cats+catf
      if(j==N){cat<-1}
    }
    if(cats==1){cat1s<-cat1s+1}
    if(cats==0){cat1s<-cat1s}
    if(catf==1){cat1f<-cat1f+1}
    if(catf==0){cat1f<-cat1f}
  }
  t1error.cal <- cat1s/sim
  N_v <- append(N_v, N)
  t1error_v <- append(t1error_v, t1error.cal)
  result2 <- cbind(N_v, t1error_v)

} # End of Type-I-error calculation


# Outputs
if (!is.null(result1) & !is.null(result2)){
   result <- merge(result1, result2, by=c("N_v"))
   final <- as.data.frame(result)

   # select the lowest/best-controlled type I error
   final$diff <- abs(final$t1error_v - t1error)
   final <- final[order(final$diff, final$t1error_v, final$power_v, final$N_v), ]
   ff <- final[1,]
   planN <- ff$N_v
   exact.power <- ff$power_v
   exact.t1 <- ff$t1error_v
   ss <- ff$n_v
   sd <- ff$sd_v
   if (prior[[1]] == 1) {method = "DIP"
   } else if (prior[[1]] == 2) {method = paste("Beta(",prior[[2]], ",", prior[[3]], ")", sep="")
   }

   z <- list(method = method, planned_sample_size = planN,
             efficacy_boundary = ps, futility_boundary = pf,
             exact_power = exact.power, exact_type_I_error = exact.t1,
             expected_sample_size = ss, expected_sample_size_std = sd)
   z
 } # End of Outputs

}
